cli/internal/fs/package_deps_hash.go

```diff
@@ -31,354 +21,352 @@ type PackageDepsOptions struct {
+	}
 
-// gitLsTree executes "git ls-tree" in a folder
-func gitLsTree(path string) (string, error) {
+	reader := handler(stdoutPipe)
+	entries, readErr := reader.ReadAll()
+	if readErr != nil {
+		return nil, fmt.Errorf("failed to read `git %s`: %w", commandName, readErr)
+	}
 
-	cmd := exec.Command("git", "ls-tree", "HEAD", "-r")
-	cmd.Dir = path
-	out, err := cmd.CombinedOutput()
-	if err != nil {
-		return "", fmt.Errorf("failed to read `git ls-tree`: %w", err)
+	waitErr := cmd.Wait()
+	if waitErr != nil {
+		return nil, fmt.Errorf("failed to read `git %s`: %w", commandName, waitErr)
 	}
-	return strings.TrimSpace(string(out)), nil
+
+	return entries, nil
 }
 
-func gitLsFiles(path AbsolutePath, patterns []string) (string, error) {
-	cmd := exec.Command("git", "ls-files", "-s", "--")
-	cmd.Args = append(cmd.Args, patterns...)
-	cmd.Dir = path.ToString()
-	out, err := cmd.CombinedOutput()
+// gitLsTree returns a map of paths to their SHA hashes starting at a particular directory
+// that are present in the `git` index at a particular revision.
+func gitLsTree(rootPath AbsolutePath) (map[turbopath.RelativeUnixPath]string, error) {
+	cmd := exec.Command(
+		"git",     // Using `git` from $PATH,
+		"ls-tree", // list the contents of the git index,
+		"-r",      // recursively,
+		"-z",      // with each file path relative to the invocation directory and \000-terminated,
+		"HEAD",    // at this specified version.
+	)
+	cmd.Dir = rootPath.ToString() // Include files only from this directory.
+
+	entries, err := runGitCommand(cmd, "ls-tree", gitoutput.NewLSTreeReader)
 	if err != nil {
-		return "", fmt.Errorf("failed to read `git ls-tree`: %w", err)
+		return nil, err
 	}
-	return strings.TrimSpace(string(out)), nil
-}
 
-// lsTreeRegex parses ls-tree output
-// A line is expected to look like:
-// 100644 blob 3451bccdc831cb43d7a70ed8e628dcf9c7f888c8    src/typings/tsd.d.ts
-// 160000 commit c5880bf5b0c6c1f2e2c43c95beeb8f0a808e8bac  rushstack
-var lsTreeRegex = regexp.MustCompile(`([0-9]{6})\s(blob|commit)\s([a-f0-9]{40})\s*(.*)`)
-
-func parseGitLsTree(output string) map[string]string {
-	changes := make(map[string]string)
-	if len(output) > 0 {
-
-		outputLines := strings.Split(output, "\n")
-		for _, line := range outputLines {
-			if len(line) > 0 {
-				matches := lsTreeRegex.MatchString(line)
-				if matches {
-					// this looks like this
-					// [["160000 commit c5880bf5b0c6c1f2e2c43c95beeb8f0a808e8bac  rushstack" "160000" "commit" "c5880bf5b0c6c1f2e2c43c95beeb8f0a808e8bac" "rushstack"]]
-					match := lsTreeRegex.FindAllStringSubmatch(line, -1)
-					if len(match[0][3]) > 0 && len(match[0][4]) > 0 {
-						hash := match[0][3]
-						filename := parseGitFilename(match[0][4])
-						changes[filename] = hash
-					}
-					// @todo error
-				}
-			}
-		}
+	output := make(map[turbopath.RelativeUnixPath]string, len(entries))
+
+	for _, entry := range entries {
+		output[turbopath.RelativeUnixPath(entry[3])] = entry[2]
```

Comment:
I just realized that the conclusion of this was in Slack, moving it here:

> I've punted on reformatting `[][]string` because it requires doing lots rearranging and reassigning at runtime in a possibly hot path.